source/simulator/stokes_matrix_free.cc

```diff
@@ -1606,6 +1674,126 @@ namespace aspect
+      // and the strain rate.
+      if (sim.newton_handler && sim.newton_handler->parameters.newton_derivative_scaling_factor!=0)
+        {
+          const double newton_derivative_scaling_factor =
+            sim.newton_handler->parameters.newton_derivative_scaling_factor;
+
+          FEValues<dim> fe_values (*sim.mapping,
+                                   sim.finite_element,
+                                   quadrature_formula,
+                                   update_values   |
+                                   update_gradients |
+                                   update_quadrature_points |
+                                   update_JxW_values);
+
+          MaterialModel::MaterialModelInputs<dim> in(fe_values.n_quadrature_points, sim.introspection.n_compositional_fields);
+          MaterialModel::MaterialModelOutputs<dim> out(fe_values.n_quadrature_points, sim.introspection.n_compositional_fields);
+          sim.newton_handler->create_material_model_outputs(out);
+#if DEAL_II_VERSION_GTE(9,3,0)
+          const unsigned int n_cells = stokes_matrix.get_matrix_free()->n_cell_batches();
+#else
+          const unsigned int n_cells = stokes_matrix.get_matrix_free()->n_macro_cells();
+#endif
+          const unsigned int n_q_points = quadrature_formula.size();
+
+          active_strain_rate_table.reinit(TableIndices<2>(n_cells, n_q_points));
+          active_viscosity_derivative_wrt_pressure_table.reinit(TableIndices<2>(n_cells, n_q_points));
+          active_viscosity_derivative_wrt_strain_rate_table.reinit(TableIndices<2>(n_cells, n_q_points));
+
+          for (unsigned int cell=0; cell<n_cells; ++cell)
+            {
+#if DEAL_II_VERSION_GTE(9,3,0)
+              const unsigned int n_components_filled = stokes_matrix.get_matrix_free()->n_active_entries_per_cell_batch(cell);
+#else
+              const unsigned int n_components_filled = stokes_matrix.get_matrix_free()->n_components_filled(cell);
+#endif
+              for (unsigned int i=0; i<n_components_filled; ++i)
+                {
+                  typename DoFHandler<dim>::active_cell_iterator FEQ_cell =
+                    stokes_matrix.get_matrix_free()->get_cell_iterator(cell,i);
+                  typename DoFHandler<dim>::active_cell_iterator sim_cell(&(sim.triangulation),
+                                                                          FEQ_cell->level(),
+                                                                          FEQ_cell->index(),
+                                                                          &(sim.dof_handler));
+
+                  fe_values.reinit(sim_cell);
+                  in.reinit(fe_values, sim_cell, sim.introspection, sim.current_linearization_point);
+
+                  sim.material_model->fill_additional_material_model_inputs(in, sim.current_linearization_point, fe_values, sim.introspection);
+                  sim.material_model->evaluate(in, out);
```

Comment:
Please flip this order. I do not know if any material relies on the order (I dont think so because `evaluate` is `const`), but still we usually first evaluate then fill additional outputs.

Reply:
I am confused. We need to fill the inputs before or not?

Reply:
You are right, I read over the `inputs` and thought they were `outputs`.

Reply:
So I don't need to flip the order here?

Reply:
The current code is correct.